### ISSUE TITLE
Multi-column Float Elements (Figures, Tables, Code Blocks, etc.)

### ISSUE THREAD

**Issue by giannissc** [typst/typst#553]
Supersedes: #137.

I want to be able to break the column rules for figures and tables to the full size of the page and I want to be able to float these elements. I believe these two must be tackled together as you can see in the [motivating examples](https://github.com/typst/typst/issues/553#issuecomment-1497070366)

**Comment by giannissc** [comment]
Using the following template:

``` typst
// The project function defines how your document looks.
// It takes your content and some metadata and formats it.
// Go ahead and customize it to your liking!
#let project(
  title: "",
  abstract: [],
  authors: (),
  date: none,
  body,
) = {
  // Set the document's basic properties.
  set document(author: authors.map(a => a.name), title: title)
  set page(numbering: "1", number-align: center)
  set text(font: "Linux Libertine", lang: "en")
  set heading(numbering: "1.1")

  // Title row.
  align(center)[
    #block(text(weight: 700, 1.75em, title))
    #v(1em, weak: true)
    #date
  ]

  // Author information.
  pad(
    top: 0.5em,
    x: 2em,
    grid(
      columns: (1fr,) * calc.min(3, authors.len()),
      gutter: 1em,
      ..authors.map(author => align(center)[
        *#author.name* \
        #author.email \
        #author.affiliation
      ]),
    ),
  )

  // Abstract.
  pad(
    x: 2em,
    top: 1em,
    bottom: 1.1em,
    align(center)[
      #heading(
        outlined: false,
        numbering: none,
        text(0.85em, smallcaps[Abstract]),
      )
      #abstract
    ],
  )

  // Main body.
  set par(justify: true)
  show: columns.with(2, gutter: 1.3em)

  body
}

#let primary_colour = rgb("#3E0C87") // vivid purple
#let max_rating = 3
#let skill(rating) = {
  let done = false
  let i = 1
  while (not done){
    let colour = rgb("#c0c0c0") // grey

    if (i <= rating){
      colour = primary_colour
    }

    box(circle(
      radius: 4pt,
      fill: colour
    ))

    if (max_rating == i){
      done = true
    } else {
      // no spacing on last
      h(2pt)
    }

    i += 1
  }

  [\ ]
}
```

**Comment by giannissc** [comment]
And the following `main.typ`:

```typst
#import "template.typ": *

// Take a look at the file `template.typ` in the file panel
// to customize this template and discover how it works.
#show: project.with(
  title: "Test Paper",
  authors: (
    (name: "John Skottis", email: "giannis.sc@gmail.com", affiliation: "The University of Edinburgh"),
  ),
  // Insert your abstract after the colon, wrapped in brackets.
  // Example: `abstract: [This is my abstract...]`
  abstract: lorem(59),
  date: "March 22, 2023",
)

// We generated the example code below so you can see how
// your document will look. Go ahead and replace it with
// your own content!

= Introduction
#lorem(60)

== In this paper
#lorem(20)

=== Contributions
#lorem(40)

= Related Work
#lorem(500)

#align(center, table(
    columns:(auto, auto, auto, auto, auto, auto, auto, auto),
    [*Column 1*], [*Column 2*], [*Multi #linebreak() Column 3*], [*Multi #linebreak() Column 4*], [*Multi #linebreak() Column 5*], [*Column 6*], [*Multi #linebreak() Column 7*], [*Column 8*],
    [Row 1], [#skill(3)], [#skill(3)], [#skill(1)], [#skill(3)], [#skill(1)], [#skill(1)], [✗],
    [Row 2], [#skill(1)], [#skill(3)], [#skill(1)], [#skill(2)], [#skill(2)], [#skill(1)], [✗],
    [Multi #linebreak() Row 3], [#skill(1)], [#skill(1)], [#skill(3)], [#skill(1)], [#skill(3)], [#skill(1)], [#sym.checkmark],
    [Multi #linebreak() Row 4], [#skill(1)], [#skill(2)], [#skill(3)], [#skill(1)], [#skill(3)], [#skill(3)], [✗],
    [Row 5], [#skill(1)], [#skill(2)], [#skill(3)], [#skill(1)], [#skill(3)], [#skill(3)], [#sym.checkmark],

))

#lorem(500)
```

**Comment by giannissc** [comment]
The rendering of the table is quite awkward
[Test Paper.pdf](https://github.com/typst/typst/files/11141133/Test.Paper.pdf)

**Comment by giannissc** [comment]
Removing the following line from `template.typ`: ` show: columns.with(2, gutter: 1.3em)` and controlling the columns from within the main function result in the following pdf:
[Test Paper-1.pdf](https://github.com/typst/typst/files/11141154/Test.Paper-1.pdf)

**Comment by giannissc** [comment]
Proposal for API change:

- Introduce `#wrap` or `#float` which will be the counterpart to `#place`. `#place` does not interrupt the flow but `#wrap` does
- The function signatures would be:
    - place(positioning, alignment, dx, dy, body)
    - wrap(positioning, alignment, dx, dy, conflict: alignment, body)
- Positioning can either be relative to page or relative to parent (existing behavior of place)

**Comment by giannissc** [comment]
Algorithm for `#wrap` elements:
1. Calculate `#wrap` element's dimensions and check if the element can fit to existing page/column
2. If it can, place `#wrap` element at correct position in page/column based on its `alignment` or `conflict` policies. Relayout all other content (except other `#wrap` elements) in current page/column.
3. If not, place `#wrap` element at correct position in the next page/column (based on its 'alignment' or `fallback` policies. Keep adding content to current page until it is filled.

**Comment by giannissc** [comment]
Conflict resolution:
- In case 2 or more wrap elements is placed in the same position in a page/column the `fallback` argument can be used to decide it position.
- It can be same as its original `alignment` in which case you place it to the next available spot (e.g. if 2 `#wrap` elements request the `top` position then the first `#wrap` element and the second to the one below it.
- It can be a different from its original `alignment`. For example if the frist `#wrap` request the `top` position then the second can request in the `fallback` the `bottom`.
   - Assuming there are no more conflicts the content is laid out again in the available space.
   - If there is another conflict move to the next page and repeat (`alignment` -> `fallback`)
   - Fallback value can be an array e.g. [alignment, alignment] if someone desires many different fallbacks.
- If no fallback value is provided move to the next page and repeat (`alignment`)

**Comment by giannissc** [comment]
Does this seem reasonable? @laurmaedje, @reknih

Also is this the best place to discuss or do you prefer discord?

I am also willing to contribute code to get this out sooner! 🙂

**Comment by giannissc** [comment]
# Motivating Examples
## No wrap (existing behavior)
Not allowing text to wrap around other elements can result in the following issue:
![Screenshot 2023-04-05 at 08-05-05 Test Paper (copy) – Typst](https://user-images.githubusercontent.com/20277283/230014812-5154d282-d64b-472b-8235-ed7dc5fc3dbe.png)

##No wrap,break column rules (global scope)
Simple allowing content to break the column rules and layout based on page will also have the same issue:
![Screenshot 2023-04-05 at 08-20-18 Test Paper – Typst](https://user-images.githubusercontent.com/20277283/230014874-7b111493-e3a4-4d0e-86a8-ce0d2013a555.png)

## Wrap (local scope) - `#wrap(scope: local, alignment: [])`
![Screenshot 2023-04-05 at 08-08-54 Test Paper (copy) – Typst](https://user-images.githubusercontent.com/20277283/230015121-8cedc6fb-f243-4fc8-9b1b-7a404fd467e9.png)

## Wrap (global scope) - `#wrap(scope: global, alignment: [])`
![Screenshot 2023-04-05 at 08-22-12 Test Paper – Typst](https://user-images.githubusercontent.com/20277283/230015683-08bef5aa-77a6-4450-8213-688ba6ef8265.png)

## Wrap (global scope) +  Alignment(Centered) - `#wrap(scope: local, alignment: [center+horizon])`
![Screenshot 2023-04-05 at 08-34-00 Test Paper – Typst](https://user-images.githubusercontent.com/20277283/230016472-8b92527d-4897-4f79-ade4-bb765edb116a.png)

## Wrap (global scope) + Conflict (Repeated) - `#wrap(scope: local, alignment: [top, top])`
![Screenshot 2023-04-05 at 08-35-25 Test Paper – Typst](https://user-images.githubusercontent.com/20277283/230016539-4dd1f08a-8b93-406c-b4c9-c49d6371d4f6.png)

## Wrap (global scope) + Conflict (Alternate) - `#wrap(scope: local, alignment: [top, bottom])`
![Screenshot 2023-04-05 at 08-37-54 Test Paper – Typst](https://user-images.githubusercontent.com/20277283/230016797-5a975dfb-3478-461e-acf2-3f4ab0a469f8.png)

**Comment by Andonome** [comment]
This is great. In fact, it'd be ideal.

**Comment by giannissc** [comment]
Do you think the primary and secondary alignments are both necessary? @Andonome

**Comment by eltos** [comment]
This is the most important missing feature to make typst as powerful as LaTeX.

**Comment by Andonome** [comment]
> Do you think the primary and secondary alignments are both necessary? @Andonome

I wouldn't say 'necessary', but I've found a great deal of flexibility in having both 'top' and 'bottom' options.

This page looks fine, as-is, but the two elements float apart when more text is added, which matches the general format of pictures-at-bottom, and tables-on-top.

![image](https://user-images.githubusercontent.com/52453904/230198361-d68b7b4a-e62e-4894-901b-9b5bce990ba1.png)

I've not found tables in the centre work very often.

Of course if anyone wants to get *really* fancy, automatic wrapping around a PNGs filled parts would be [punk-as-all-heck](https://external-content.duckduckgo.com/iu/?u=http%3A%2F%2Fcdn.escapistmagazine.com%2Fmedia%2Fglobal%2Fimages%2Flibrary%2Fderiv%2F773%2F773273.jpg&f=1&nofb=1&ipt=e883cc1ae984d66072bff2745f0e2fbcb1f147264d4e2f6b3cca294abfc4a10f&ipo=images).

**Comment by Andonome** [comment]
> This is the most important missing feature to make typst as powerful as LaTeX.

LaTeX doesn't have this ability. There is no reliable text wrapping in a multicols environment.  It's a big factor which stops it having reliable output - every compile, every page needs to be checked, images moved, et c.

**Comment by giannissc** [comment]
@Andonome I think is wrap is to be implemented it should be implemented with all these usecases in mind. Another usecase that comes to mind is that you have a central circular figure and that text should wrap around it as well. You examples are genuinely useful. Thanks!

**Comment by giannissc** [comment]
In you table and figure example would you want to force text to move to the next page or would it be ok for it to render in-between the two elements?

**Comment by Andonome** [comment]
> @Andonome I think is wrap is to be implemented it should be implemented with all these usecases in mind. Another usecase that comes to mind is that you have a central circular figure and that text should wrap around it as well. You examples are genuinely useful. Thanks!

Would a showcase be useful? I've had a bunch of ugly results with difficult elements.



> In you table and figure example would you want to force text to move to the next page or would it be ok for it to render in-between the two elements?

It depends on the results. Either could be fine, but what's sorely lacking is the ability to return to the last page for a do-over. LaTeX can't correct previous pages, so you can get these ugly dangling single-lines of text.

I've given headers a bunch of `\needspace{` commands to stop them going at the end of a page, but that's left this dangling, when the next couple of `\paragraph` statements (with **Journeys**) could have fit between these two images fine.

Of course here it's blocked by the **General Tempo Chart**, so that should have been moved to after the next paragraph.

![image](https://user-images.githubusercontent.com/52453904/230340272-2b5a5234-a307-427d-8b06-c86449f7482a.png)

In a lot of cases, it's far more important to say 'place this element somewhere before the next subsection' than making a statement about the position.

Here's another table that could have been better placed by floating anywhere else before the `\paragraph{`.

![image](https://user-images.githubusercontent.com/52453904/230344276-71ebb3bc-7b97-4613-a684-edd388bfe360.png)

I don't know if there's a non-intensive way to have typst go through a bunch of options involving multiple pages, but maybe there could be a second processing option where the user understands it'll take extra time to process, but the results will make sense. Atm a full LaTeX process can take 20 minutes - that's quite doable for big pushes, but less okay when the results need fiddled with before publishing a new version.

**Comment by giannissc** [comment]
Ok I just had another look and your usecase in incredibly useful! I might need to tweak the `scope` argument with other options such as `paragraph`, `section`, etc. for more fine-grained control!

**Comment by giannissc** [comment]
> It depends on the results. Either could be fine, but what's sorely lacking is the ability to return to the last page for a do-over. LaTeX can't correct previous pages, so you can get these ugly dangling single-lines of text.

I am not sure I fully understand this point. Is this something that the algorithm is lacking or is it something that the existing Latex API cannot express. If it is the latter I will make a note on the algorithm implementation portion of the proposal otherwise I will make a note on the motivation and API section as an additional user intent

**Comment by giannissc** [comment]
Proposal v3:
[Multi-column Float Proposal-3.pdf](https://github.com/typst/typst/files/11170510/Multi-column.Float.Proposal-3.pdf)

**Comment by Andonome** [comment]
> Proposal v4: [Multi-column Float Proposal-4.pdf](https://github.com/typst/typst/files/11170593/Multi-column.Float.Proposal-4.pdf)

This all looks fantastic.

In 1.6:

> Giving an option to
 disable horizontal wrapping would alleviate the issue.

It might be better to disable this by default. Nobody wants to see 2-words crawling down the side of a table. The table should be centred automatically.

In fact in general, the default being 'typst chooses' is always good.  Take for example the first example table.  It's clearly best presented image-wide, but that won't always be obvious until the page has processed.  If the floating default behaviour was for typst to choose how it floats (and only request the user states if it should be in the same section or page as something else) then it'd really help in producing unattended document production.

**Comment by 0x7CFE** [comment]
I can suggest reading [the Lillypond essay](https://lilypond.org/doc/v2.23/Documentation/essay/music-engraving) about problems of automated music engraving. I believe, a lot of ideas are common to any complex layout systems and potentially can be applied to typst as well.

tldr: Music notation is much more complicated than one can imagine. Naive approach to notation layout based on hierarchy of containers (page→staff→lines→notes etc) simply does not work. Once in a while you will face some quirky notation like multi-staff slurs or beams, and things will get complicated. In the end, this will result in a lot of ad-hoc code, or the result will not be visually pleasing. Lillypond uses completely different approach to layout.

**Comment by granthamtaylor** [comment]
I just wanted to add my +1 here.

I'm trying to write up a whitepaper in IEEE format and floating tables / figures are not possible AFAIK.

**Comment by KronosTheLate** [comment]
It would be lovely to see this feature implemented.

**Comment by coljac** [comment]
+1. As far as I can tell, this is the only missing feature stopping me completing a nearly perfect template for my usual research journal (and then hassling the editors about it). Having figures and tables spanning two columns is a must.

**Comment by Andonome** [comment]
Just thought I'd add some god-tier image floating challenges.


![Image taking up 2 out of 3 columns, at top](https://github.com/typst/typst/assets/52453904/33296448-045d-4c80-9bf3-4e4d9a30a271)

The ideal command here would be something like `#image[ columns = 2, image.png]`.

![Image overlapped by text](https://github.com/typst/typst/assets/52453904/e189ad93-9491-4162-bc00-65002a792978)

I don't how much of a 'badness' rating typst has under the hood, but perhaps an image could be covered when needed, as above.

![text-wrap](https://github.com/typst/typst/assets/52453904/ee3a6e50-fb7d-4639-9acf-78f4b2773849)

Clearly only available for .png or something with transparency, but automatic text-wrapping would be the bees' knees.

**Comment by jaanli** [comment]
+1 need this as well!

**Comment by rtbs-dev** [comment]
Also needing this for conference/journal submissions. E.g. summary tables, larger figures (a la small-multiples), etc.

**Comment by danaugrs** [comment]
Any updates on this? I need a way to span a table over two columns for a research paper.

**Comment by laurmaedje** [comment]
@danaugrs Unfortunately, there has been no progress on this lately.

**Comment by let-me-cook** [comment]
Is this feature being considered for version 1.0 of Typst?

**Comment by laurmaedje** [comment]
> Is this feature being considered for version 1.0 of Typst?

Definitely. But note that 1.0 is still a ways off.

**Comment by Polydynamical** [comment]
We need this feature Typst Devs :cry:

**Comment by ntjess** [comment]
A solution in the meantime: https://github.com/ntjess/wrap-it
PR is here: https://github.com/typst/packages/pull/357

It handles the common cases (think of it like a less-finicky version of `wrapfig`), but there are still some fundamental limitations that can't be solved in typst
![image](https://github.com/typst/typst/assets/23620506/c04cfd1a-9dec-4bf2-84b2-0d994c93d141)

**Comment by denkspuren** [comment]
I'm impressed by your solution Nathan! Great work!

**Comment by davidleejy** [comment]
> A solution in the meantime: https://github.com/ntjess/wrap-it PR is here: [typst/packages#357](https://github.com/typst/packages/pull/357)
> 
> It handles the common cases (think of it like a less-finicky version of `wrapfig`), but there are still some fundamental limitations that can't be solved in typst

This looks pretty neat. Does anyone know whether it's possible to position a table spanning the width of the page in typical Typst 2-column layout with this package?

**Comment by danaugrs** [comment]
> > A solution in the meantime: https://github.com/ntjess/wrap-it PR is here: [typst/packages#357](https://github.com/typst/packages/pull/357)
> > It handles the common cases (think of it like a less-finicky version of `wrapfig`), but there are still some fundamental limitations that can't be solved in typst
> 
> This looks pretty neat. Does anyone know whether it's possible to position a table spanning the width of the page in typical Typst 2-column layout with this package?

Doesn't seem like it's possible - at least I wasn't able to. Maybe @ntjess knows a workaround?

**Comment by RaulDurand** [comment]
Im not sure if it was already discussed elsewhere but I believe that the ability to place multicolumn floats should be a feature of `columns` (reflowing the text as needed) instead of a temporary break of the `columns` element.

**Comment by the-busy-beaver** [comment]
i encountered the same _issue_ whilst using the [charged-iee](https://typst.app/universe/package/charged-ieee/) template, i wasn't able to put a figure containing a large / wide image at the top, spanning the two columns :thinking: 

having this feature or a workaround would be really great :innocent:

**Comment by Vivdaddy** [comment]
Any idea when this feature will be added?

**Comment by laurmaedje** [comment]
@Vivdaddy We plan to take a closer look at this for 0.12, but I can't make any promises.

**Comment by danaugrs** [comment]
It seems like spanning a figure across all columns should be relatively straightforward to implement and it would fix 90% of all multi-column issues. Specially if the figure is at the top or bottom of a page. It would be as if a page with such a figure were a little shorter and the page break happened a bit before or after the actual page break in order to make space for the figure. I know that if this tiny feature were implemented it would prevent me from temporarily switching back to LaTeX, and it seems a lot of people are also anxiously waiting for this (over 100 users if we go by GitHub reactions, but I suspect there are many more). Please let us know when you have updates! :heart:

**Comment by Enivex** [comment]
> It seems like spanning a figure across all columns should be relatively straightforward to implement and it would fix 90% of all multi-column issues. Specially if the figure is at the top or bottom of a page. It would be as if a page with such a figure were a little shorter and the page break happened a bit before or after the actual page break in order to make space for the figure. I know that if this feature were implemented it would prevent me from temporarily switching back to LaTeX, and it seems a lot of people are also anxiously waiting for this (over 100 users if we go by GitHub reactions, but I suspect there are many more). Please let us know when you have updates! :heart: 

I suspect that if it was straightforward the issue wouldn't still be open.

**Comment by danaugrs** [comment]
> I suspect that if it was straightforward the issue wouldn't still be open.

This issue is about arbitrary multi-column float elements, which is a complex problem, specially if it involves arbitrary-shape text wrapping as in [the last image in this comment](https://github.com/typst/typst/issues/553#issuecomment-1560215098). My objective was to point out that there is a subset of this problem space i.e. _spanning an element across **all** columns of a page, specially at the top or bottom of a page_ (with no arbitrary-shape text wrapping), which is simpler to implement, and would solve the majority of use cases. I think any attempt at a solution should first focus on this.

**Comment by Polydynamical** [comment]
@laurmaedje Any progress on this issue? :)

**Comment by RaulDurand** [comment]
I believe that the lack of this feature is the main limitation to using Typst for writing manuscripts for scientific journals.

**Comment by kescobo** [comment]
> is the main limitation

It depends on the field - almost nothing I submit to (in biology) uses multi-column layouts, but I need line numbers (happily that's being worked on #352 )

**Comment by danaugrs** [comment]
I think the ideal solution, from a user's perspective, would involve adding a single parameter, tentatively called `colspan`, to non-breakable block elements, which would specify how many columns that element should span. Specifying a `colspan` of more columns than exist in the current context would result in an error. I think such a parameter, coupled with the existing usage of `placement: "auto" | "top| "bottom"`, would provide complete freedom to place multi-column blocks anywhere as desired.

**Comment by granthamtaylor** [comment]
> Definitely true for informatics and computer sciences where the IEEE format is quite common. Some stuff is just to large to fit in a single half-page column meaningfullly.

This what lead to me following this issue last year. I am in ML/AI where IEEE is standard. At the same time, I have several large diagrams that I wish to display inline. There is no way I could possibly adapt my diagrams to fit in two-column layout.

**Comment by jounathaen** [comment]
I'm not a mod, but I think that there are now enough "_This feature is stopping me from using Typst_"/"_This feature is missing for XYZ support_" comments in this issue. The Devs are aware of this issue, and it is as-of-now the third most upvoted issue. So instead of adding more comments of the same, I'd suggest using the emoticons to express your support for this feature. 
(And the selfish remark: Then people that have subscribed to this issue will not get spammed with non-progress messages)

**Comment by NeutrinoLiu** [comment]
<img width="1579" alt="Screenshot 2024-08-29 at 12 16 20 AM" src="https://github.com/user-attachments/assets/05580089-bcdd-4f96-b0ba-2053be303bf2">
in case anyone who is still waiting for the new feature, here is my bypass (i do really like typst and would never go back to latex):
add your figure to the first column and wrap it with a place function, so that it appears in the middle of two cols. Then at the top of second column, you place another empty rect, so that contents on the right side will align with left side.
`#place(
  top,
  dx: 105%,
  float: true,
  figure(
    table(
    align: center,
    columns: (auto, auto, auto, auto, 1fr),
    [Primitive], [PSNR], [FPS], [Accessibility], [Storage],
  
    [3DGS @3dgs], [high], [high (>100)], [relative easy], [middle],
    [NeRF @nerf], [high], [low (\<10)], [relative easy], [low],
    [point cloud], [depends], [middle], [easy], [high],
    [mesh], [depends], [high], [poor], [middle]
    ),
    caption: [Comparison between different 3D representation]
    ),
)
#lorem(500)
#place(
  top,
  float: true,
  rect(height: 101pt, fill: none, stroke: none)
)`
you might want to adjust the location of the second place and the rect height so that visually it looks like that the table spans two columns.

**Comment by spped2000** [comment]
`#place(
  top,
  dx: 70%, // Horizontal centering
  float: true, // Enable floating to allow automatic positioning
  clearance: 0pt,
  figure(
    place(
      auto,
      dx: 0%, // Center the image horizontally within its container
      float: true, // Allow image floating
      clearance: 0pt,
      image("A2.png", width: 200%)
    ),
    caption: [Visualization of Zero-Shot, Few-Shot model.] // Centered under the image
  ),
)
#place(
  top,
  float: true, // Floating to ensure proper positioning
  rect(height: 5pt, fill: none, stroke: none)
)
#lorem(500)
#place(
  top,
  float: true, // Floating to ensure proper positioning
  rect(height: 290pt, fill: none, stroke: none)
)`
![Screenshot 2024-09-19 102439](https://github.com/user-attachments/assets/8fd3fcbc-3193-4878-ac60-4beb3eb20dd9)

**Comment by the-busy-beaver** [comment]
@NeutrinoLiu and @spped2000 :wave: 

so i've tried your examples and they did not work on my side...
could you share a complete repro of your examples? that would be amazing

### my repro
i have installed Typst today with
```shell
cargo +1.77 install --git https://github.com/typst/typst --locked typst-cli
```
and i get the following `typst --version`
```
typst 0.11.0 (8c813cb7)
```

my Typst source is a single file, called `main.typ` and which contains
```typst
#show: rest => columns(2, rest)

#place(
 top,
 dx: 105%,
 float: true,
 clearance: 0pt,
 figure(
   place(
     auto,
     dx: -40%,
     float: true,
     clearance: -8pt,
     figure(
       table(
         align: center,
         columns: (80pt, 80pt, 80pt, 80pt, 80pt),
         [Primitive], [Rendering Quality], [Rendering Speed], [Accessibility], [Storage],
         [3DGS 3dgs], [high], [high (>100fps)], [relatively easy], [middle],
         [NeRF nerf], [high], [low (\<10fps)], [relatively easy], [low],
         [point cloud], [depends], [middle], [easy], [high],
         [mesh], [depends], [high], [poor], [middle]
       )
     ),
   ),
   caption: [Comparison between different 3D representation]
 ),
)
#lorem(500)
#place(
 top,
 float: true,
 rect(height: 87pt, fill: none, stroke: none)
)
```

then i compile this with
```typst
typst compile main.typ main.pdf
```
and here is what i get :confused: 
![Screenshot from 2024-09-23 14-09-50](https://github.com/user-attachments/assets/63727904-dda7-4fbb-bf0b-3a8f44ddcd94)

**Comment by laurmaedje** [comment]
Good news: https://github.com/typst/typst/pull/5017

**Comment by sschuldenzucker** [comment]
Hey guys, I'm probably missing something but can I use the new layout engine to wrap text around figures? [The wrap-it package](https://typst.app/universe/package/wrap-it) says that it's only relevant as long as this issue isn't resolved. Since it is now, does this mean I can actually wrap text around figures? (and how?) Or is it still not possible and the package documentation should be updated?

**Comment by laurmaedje** [comment]
@sschuldenzucker Text wrapping around figures at the side was not part of the 0.12 work on floating placement. This subfeature was also proposed in this issue, but it was too large of an issue to be actionable. Side-wrapping is now tracked by https://github.com/typst/typst/issues/5181.

**Comment by AstroAure** [comment]
Hello !
I'm re-opening this discussion since I can't manage to place my figures the ways I would like. Basically, I want my figures to behave like [Wrap (local scope)](https://github.com/typst/typst/issues/553#issuecomment-1497070366) shown above.
With the ``placement`` option, this leads to the following logic:

* If the figure has the space to stay on the same page/column: use ``placement: none``
<img width="300" height="450" alt="Image" src="https://github.com/user-attachments/assets/4a945c7b-292f-4c48-af92-e8f62524fc7d" />

* If the figure is pushed to a new page because it doesn't have enough space: use ``placement: top``
<img width="300" height="450" alt="Image" src="https://github.com/user-attachments/assets/47ac0f0e-ecd6-4598-9ae3-6fb997967907" />

This ensures that the image is placed at the correct spot in the text if it it has the space for it, and otherwise gets placed at the top of the very next page/column, while continuing the text before to not leave blank space.
Is there a way to achieve this behavior in the current version of Typst ? Otherwise, would this be possible to add in future versions ?
Thank you !

**Comment by kescobo** [comment]
@AstroAure I don't think it makes sense to extend this old issue since the discussion has moved on. If you think this may be a bug in the implementation from v0.12, you can open a new issue (or better yet hop on discord and ask for help in case there's a good solution), or if it's relevant to #5181 you could comment there.
